linkerd/identity/src/imp/openssl.rs

```diff
@@ -0,0 +1,345 @@
+            InvalidCrt::General(err) => err.source(),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CrtKey {
+    id: LocalId,
+    expiry: SystemTime,
+    client_config: Arc<ClientConfig>,
+    server_config: Arc<ServerConfig>,
+}
+
+// === CrtKey ===
+impl CrtKey {
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+
+    pub fn expiry(&self) -> SystemTime {
+        self.expiry
+    }
+
+    pub fn id(&self) -> &LocalId {
+        &self.id
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        self.client_config.clone()
+    }
+
+    pub fn server_config(&self) -> Arc<ServerConfig> {
+        self.server_config.clone()
+    }
+}
+
+impl fmt::Debug for CrtKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_struct("CrtKey")
+            .field("id", &self.id)
+            .field("expiry", &self.expiry)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Crt {
+    pub(crate) id: LocalId,
+    expiry: SystemTime,
+    pub cert: X509,
+    pub chain: Vec<X509>,
+}
+
+impl Crt {
+    pub fn new(
+        id: LocalId,
+        leaf: Vec<u8>,
+        intermediates: Vec<Vec<u8>>,
+        expiry: SystemTime,
+    ) -> Self {
+        let mut chain = Vec::with_capacity(intermediates.len() + 1);
+        let cert = X509::from_der(&leaf).unwrap();
+        chain.extend(
+            intermediates
+                .into_iter()
+                .map(|crt| X509::from_der(&crt).unwrap()),
+        );
+
+        Self {
+            id,
+            cert,
+            chain,
+            expiry,
+        }
+    }
+
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+}
+
+#[derive(Clone)]
+pub struct ClientConfig {
+    pub root_certs: Arc<X509Store>,
+    pub key: Option<Arc<Key>>,
+    pub cert: Option<Arc<Crt>>,
+    pub protocols: Arc<Vec<Vec<u8>>>,
+}
+
+impl fmt::Debug for ClientConfig {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&self.protocols, f)
+    }
+}
+
+impl ClientConfig {
+    pub fn new(
+        protocols: Vec<Vec<u8>>,
+        root_certs: Arc<X509Store>,
+        cert: Option<Crt>,
+        key: Option<Key>,
+    ) -> Self {
+        Self {
+            root_certs,
+            protocols: Arc::new(protocols),
+            key: key.map(Arc::new),
+            cert: cert.map(Arc::new),
+        }
+    }
+    pub fn empty() -> Self {
+        ClientConfig::new(
+            Vec::new(),
+            Arc::new(X509StoreBuilder::new().unwrap().build()),
+            None,
+            None,
+        )
+    }
+
+    pub fn set_protocols(&mut self, protocols: Vec<Vec<u8>>) {
+        self.protocols = Arc::new(protocols)
+    }
+}
+
+#[derive(Clone)]
+pub struct ServerConfig {
+    pub root_certs: Arc<X509Store>,
+    pub key: Option<Arc<Key>>,
+    pub cert: Option<Arc<Crt>>,
+    pub alpn_protocols: Arc<Vec<Vec<u8>>>,
```

Comment:
these shouldn't need `pub`
```suggestion
    root_certs: Arc<X509Store>,
    key: Option<Arc<Key>>,
    cert: Option<Arc<Crt>>,
    alpn_protocols: Arc<Vec<Vec<u8>>>,
```